[14.0] Add warning banner when payment order is not SEPA

alexis-via opened the pull request:
The field 'sepa' on account.payment.order is only display for SEPA payment methods.
If the option "show warning if not SEPA" is enabled on the payment method, a warning banner is now displayed on payment orders with a SEPA payment method which are not SEPA:

![sepa_warn_banner](https://user-images.githubusercontent.com/1157917/221707200-7d9eefd5-e4b1-4a7a-9199-22f0cb938c95.png)

alexis-via committed:
Add warning when payment order is not SEPA

The field 'sepa' on account.payment.order is only display for SEPA
payment methods.
If the option "show warning if not SEPA" is enabled on the payment
method, a warning banner is now displayed on payment orders with a SEPA
payment method which are not SEPA.

alexis-via commented:
Tests are now green!

pedrobaeza commented on account_banking_pain_base/models/account_payment_order.py:
```diff
@@ -105,6 +111,8 @@ def _sepa_iban_prefix_list(self):
         ]
 
     @api.depends(
+        "payment_mode_id.payment_method_id.pain_version",
```
I don't think you need this convoluted depends. Only `payment_mode_id`. If not, changing these fields in payment methods will trigger the recomputation of all payment orders, which is not performant and even not convenient.

alexis-via replied:
done

pedrobaeza commented on account_banking_pain_base/models/account_payment_order.py:
```diff
@@ -24,7 +24,13 @@
+    sepa_payment_method = fields.Boolean(
+        compute="_compute_sepa",
+        string="SEPA Payment Method",
+        store=True,
```
Please don't store these fields that are only for showing a banner, and that will be triggered on module update, making it slower.

alexis-via replied:
OK. So I removed the "sepa" field from the account.payment.order tree view, because it is a bad practise to have a non-stored field in a tree view.

alexis-via commented:
@pedrobaeza I took into account all your remarks.

alexis-via commented:
I'll forward-port it to v16 once all modules are merged (this PR impacts several different modules, some of which are not merged yet)

pedrobaeza commented:
Please fw-port it to v15 too. I'll promise to review and fast-track it. Deal?

alexis-via committed:
account_payment_order: re-implement allow_blocked

alexis-via commented:
I also rewrote the code for allow_blocked, which was a mess: there was no solution to block when push payment lines from an invoice. The allow_blocked configuration is now on the payment mode and applies both to push and pull mode.

pedrobaeza commented on account_payment_order/wizard/account_payment_line_create.py:
```diff
@@ -77,8 +76,8 @@ def _prepare_move_line_domain(self):
             domain += [("move_id.state", "=", "posted")]
-        if not self.allow_blocked:
-            domain += [("blocked", "!=", True)]
+        if not self.order_id.allow_blocked:
```
Shouldn't this be `if not self.order_id.payment_mode_id.allow_blocked:`?

alexis-via replied:
You're right. Fixed.

pedrobaeza commented on account_banking_pain_base/views/account_payment_method.xml:
```diff
@@ -10,10 +10,15 @@
         <field name="arch" type="xml">
             <field name="payment_type" position="after">
                 <field name="pain_version" />
+
```
Please avoid this introduced empty line

alexis-via replied:
fixed

pedrobaeza commented on account_banking_pain_base/models/account_payment_method.py:
```diff
@@ -16,6 +16,7 @@ class AccountPaymentMethod(models.Model):
         "the corresponding unaccented character, so that only ASCII "
         "characters are used in the generated PAIN file.",
     )
+    warn_not_sepa = fields.Boolean(string="Warn If Not SEPA")
```
I think the warning shouldn't be selectable. If the payment method is SEPA, then show the warning.

alexis-via replied:
I made it an option because I thought that this module may be used in non-sepa countries to produce ISO2022 files. And those users maybe upset to have the warning banner all the time.

pedrobaeza replied:
But the banner should only be shown when the payment method is SEPA. I think you have already checked that on your compute. No problem thus on being "mandatory". It's not going to appear for non SEPA orders.

pedrobaeza replied:
This is not attended.

alexis-via replied:
Let me re-explain my point of view in more details. The modules account_banking_sepa_credit_transfer and account_banking_sepa_direct_debit produce ISO20022 files that can be SEPA or not SEPA. We could have called those modules account_banking_iso20022_credit_transfer and account_banking_iso20022_direct_debit, but we chose to call them "sepa"... because, from our point of view of EU users, we use it mostly for SEPA payments. Same for the "code" of the payment method.
But, maybe some users outside of EU use those module to generate non-SEPA ISO20022 files. And they may not want to always have the warning banner. I must say I have no idea if the modules account_banking_sepa_xxx are used by companies outside EU. But I know that ISO20022 is a world-wide standard ; it is not just designed for EU/SEPA.

pedrobaeza commented on account_banking_pain_base/models/account_payment_order.py:
```diff
@@ -115,30 +121,41 @@ def _compute_sepa(self):
-                    sepa = False
-                    break
-                if pline.partner_bank_id.acc_type != "iban":
+            sepa_payment_method = False
```
For making the code more readable, since v13 you can assign directly to the field the variable without performance penalty, and also assign for the whole recordset.

alexis-via committed:
account_payment_order: show payment_reference

The field payment_reference of account.payment is important because it
has the label of the wire transfer. In the 'account' module,
that field is not displayed in tree view nor in form view of
account.payment. This commit fixes this.

account_banking_pain_base: remove empty line introduced by mistake

pedrobaeza commented on account_payment_order/wizard/account_payment_line_create.py:
```diff
@@ -26,7 +26,6 @@ class AccountPaymentLineCreate(models.TransientModel):
         selection=[("posted", "All Posted Entries"), ("all", "All Entries")],
         string="Target Moves",
     )
-    allow_blocked = fields.Boolean(string="Allow Litigation Move Lines")
```
Why removing this? It's useful.

alexis-via replied:
Because it is now a property of the payment mode, not a question you ask to the use in this wizard. The user can use the button on the invoice to push it on the payment order and skip the wizard.

alexis-via committed:
[FIX] account_payment_order: crash in wizard to add payment lines

pedrobaeza reviewed (changes requested):
Apart for not being all the comments attended, this is introducing other undesired changes like the move of the allow_blocked to the payment mode. Please only one change per PR for being discussed separately.

alexis-via committed:
allow_blocked: also check upon confirmation of payment order

alexis-via commented:
@pedrobaeza I made separate PRs in v16. We can focus on v16 now.
